src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -2025,146 +1982,188 @@ else if (compareAndSetCtl(c, c) && casRunState(e, e | STOP)) {
+                    return;
+                working = false;                          // reactivated
+            }
         }
     }
 
     /**
-     * Scans for and if found executes top-level tasks: Tries to poll
-     * each queue starting at initial index with random stride,
-     * returning next scan window and retry indicator.
+     * Deactivates and if necessary awaits signal or termination.
      *
-     * @param w caller's WorkQueue
-     * @param window up to three queue indices
+     * @param w the worker
      * @param r random seed
-     * @return the next window to use, with RESCAN set for rescan
+     * @param phase current phase
+     * @return current phase, with IDLE set if worker should exit
      */
-    private long scan(WorkQueue w, long window, int r) {
-        WorkQueue[] qs = queues;
-        int n = (qs == null) ? 0 : qs.length, step = (r << 1) | 1;
-        long next = window & ~RESCAN;
-        outer: for (int i = (short)window, l = n; l > 0; --l, i += step) {
-            int j, cap; WorkQueue q; ForkJoinTask<?>[] a;
-            if ((q = qs[j = i & SMASK & (n - 1)]) != null &&
-                (a = q.array) != null && (cap = a.length) > 0) {
-                for (int b = q.base;;) {
-                    int nb = b + 1, nk = nb & (cap - 1), k;
-                    ForkJoinTask<?> t = a[k = b & (cap - 1)];
-                    U.loadFence();                // re-read b and t
-                    if (b == (b = q.base)) {      // else inconsistent; retry
-                        Object o;
-                        if (t == null)
-                            o = a[k];
-                        else if (t == (o = U.compareAndExchangeReference(
-                                           a, slotOffset(k), t, null))) {
-                            q.updateBase(nb);
-                            next = RESCAN | ((window << 16) & HMASK) | j;
-                            if (window != next && a[nk] != null)
-                                signalWork(a, nk); // limit propagation
-                            if (w != null)        // always true
-                                w.topLevelExec(t, q, j);
-                            break outer;
-                        }
-                        if (o == null) {
-                            if (next >= 0L && a[nk] != null)
-                                next |= RESCAN;
-                            break;
+    private int deactivate(WorkQueue w, int r, int phase) {
+        int p = phase | IDLE, active = phase + (IDLE << 1);
+        if (w != null) {                         // always true
+            w.phase = p;
+            long pc = ctl, qc;
+            for (;;) {                           // try to enqueue
+                w.stackPred = (int)pc;           // set ctl stack link
+                qc = (active & LMASK) | ((pc - RC_UNIT) & UMASK);
+                if (pc == (pc = compareAndExchangeCtl(pc, qc)))
```

Comment:
If I'm reading this right you might want to add a comment :)

```suggestion
                if (pc == (pc = compareAndExchangeCtl(pc, qc))) // consistent
```

Reply:
thanks, done.